emma-spark/src/main/scala/org/emmalanguage/api/SparkDataset.scala

```diff
@@ -116,6 +117,16 @@ class SparkDataset[A: Meta] private[api](@transient private[emmalanguage] val re
     rep.collect()
 
   // -----------------------------------------------------
+  // Pre-defined folds
+  // -----------------------------------------------------
+
+  override def find(p: A => Boolean): Option[A] =
+    Option(fold[Tuple1[A]](Tuple1(null.asInstanceOf[A]))(
+      x => if (p(x)) Tuple1(x) else Tuple1(null.asInstanceOf[A]),
+      (x, y) => if (x._1 != null) x else y
+    )._1)
```

Comment:
Later on fold fusion might cause a problem, but for now it'll have to do.

On a more general note, there remains the unsolved problem of different backends supporting different data types. Which is made even more difficult by subtle inconsistencies like the `Option` vs `null` case, the general lack of specification and hard to debug errors popping up at runtime rather than compile time, all of this varying across versions.

It seems hopeless to reach any kind of consensus about user defined data types.